pkg/controller/sparkapplication/controller.go

```diff
@@ -258,91 +259,88 @@ func (c *Controller) processNextItem() bool {
 	return true
 }
 
-// Helper data structure to encapsulate current state of the driver pod.
-type driverState struct {
-	podName            string         // Name of the driver pod.
-	sparkApplicationID string         // Spark application ID.
-	nodeName           string         // Name of the node the driver pod runs on.
-	podPhase           apiv1.PodPhase // Driver pod phase.
-	completionTime     metav1.Time    // Time the driver completes.
-	err                error          // Error if any found.
-
-}
-
 func (c *Controller) getAppPods(app *v1beta1.SparkApplication) ([]*apiv1.Pod, error) {
 	// Fetch all the pods for the current run of the application.
 	selector := labels.SelectorFromSet(labels.Set(getResourceLabels(app)))
 	pods, err := c.podLister.Pods(app.Namespace).List(selector)
 	if err != nil {
 		return nil, fmt.Errorf("failed to get pods for SparkApplication %s/%s: %v", app.Namespace, app.Name, err)
 	}
+
 	return pods, nil
 }
 
+// Fetch driver pod from APIServer.
+func (c *Controller) getDriverPodFromApiServer(app *v1beta1.SparkApplication) (*apiv1.Pod, error) {
+	selector := labels.SelectorFromSet(map[string]string{config.SparkAppNameLabel: app.Name, config.SparkRoleLabel: config.SparkDriverRole})
+	listOptions := metav1.ListOptions{LabelSelector: selector.String()}
+	podsList, err := c.kubeClient.CoreV1().Pods(app.Namespace).List(listOptions)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get driver pod for SparkApplication from APIServer %s/%s: %v", app.Namespace, app.Name, err)
+	} else if podsList.Size() == 1 {
+		return &podsList.Items[0], nil
+	}
+	return nil, nil
+}
+
 // getAndUpdateDriverState finds the driver pod of the application
 // and updates the driver state based on the current phase of the pod.
 func (c *Controller) getAndUpdateDriverState(app *v1beta1.SparkApplication) error {
 	pods, err := c.getAppPods(app)
```

Comment:
we should be getting just the driver, this code seems unnecessarily complex. I would do a lookup in the informer cache and if not found go to etcD

Reply:
This does sound better if we can just do a get on the driverpod name rather than a list/filter using labels but there is still a edge-case where if the user updates the driver podname in the SparkApplication, we can miss terminating the old driver in such a case.

Reply:
In this case, how about using the driver pod name in the spec, and if it's not set, use the recorded driver pod name in status?

Reply:
If a user updates the spec to update the driver pod name, we will miss terminating the old driver pod (because the name has now changed). I think this is a slight edge-case and just want to make sure we are aware of this and ok with this.